src/extensions/diagnose.ts

```diff
@@ -0,0 +1,62 @@
+import { logPrefix } from "./extensionsHelper";
+import { getProjectNumber } from "../getProjectNumber";
+import * as utils from "../utils";
+import * as resourceManager from "../gcp/resourceManager";
+
+const SERVICE_AGENT_ROLE = "roles/firebasemods.serviceAgent";
+
+/**
+ * Diagnoses and optionally fixes known issues with project configuration, ex. missing Extensions Service Agent permissions.
+ * @param projectId ID of the project we're querying
+ * @param fix Whether identified issues should be automatically fixed.
+ */
+export async function diagnose(projectId: string, fix: boolean): Promise<boolean> {
+  const projectNumber = await getProjectNumber({ projectId });
+  const firexSaProjectId = utils.envOverride(
+    "FIREBASE_EXTENSIONS_SA_PROJECT_ID",
+    "gcp-sa-firebasemods"
+  );
+
+  const saEmail = `service-${projectNumber}@${firexSaProjectId}.iam.gserviceaccount.com`;
+
+  utils.logLabeledBullet(logPrefix, "Checking project IAM policy...");
+
+  const policy = await resourceManager.getIamPolicy(projectId);
+  let foundP4saInPolicy = false;
+  policy.bindings.forEach((b) => {
```

Comment:
Prefer using `for... of...` instead of using `forEach()`: go/tsstyle#iterating-containers